# Fine-tuning DistilBERT for Named Entity Recognition (NER)

In this notebook I fine-tuned a pre-trained Transformer model for a token-level task: **Named Entity Recognition (NER)**.  
Unlike simple text classification, where the model predicts a single label per input, in NER we assign a label to **each token** in the sentence (e.g. `B-PER`, `I-LOC`, `O`).

We use:
- The **CoNLL-2003** https://huggingface.co/datasets/lhoestq/conll2003 dataset, a standard benchmark for NER in English newswire.
- **DistilBERT** https://huggingface.co/distilbert/distilbert-base-uncased as a base model, which is a compressed version of BERT with fewer parameters and faster inference.
- The HuggingFace 🤗 ecosystem (`datasets`, `transformers`, `Trainer`) to handle data, model, training loop, and evaluation.

The goal is to:
1. Prepare the CoNLL-2003 dataset for token classification.
2. Fine-tune DistilBERT on the training set.
3. Evaluate the model using sequence-level metrics (F1, precision, recall) suitable for NER.
4. Inspect qualitative examples of the model predictions.


The project can be executed by this link: https://colab.research.google.com/drive/1EtS2tQ7rTiQmjJuZITH_n8ooW3Yn1Kwi?usp=sharing

## 2. Setup

We start by installing and importing the required libraries:
- `transformers` for the model, tokenizer, and training utilities.
- `datasets` for loading and processing the CoNLL-2003 dataset.
- `seqeval` for NER-specific evaluation metrics.


## Check for GPU
Make sure you have your GPU available. You will need it.

In [ ]:
import torch

# Check for GPU availability
if torch.cuda.is_available():
    print(f"GPU detected: {torch.cuda.get_device_name(0)}")
else:
    print("No GPU detected.")
    print("If you are using Google Colab, please go to 'Runtime' > 'Change runtime type' and select 'GPU' as the hardware accelerator.")

GPU detected: Tesla T4


## Installation of libraries

We use the following libraries: `transformers`, `datasets`, and `evaluate`, plus `seqeval` for NER metrics.


In [ ]:
!pip install -q transformers datasets seqeval accelerate

In [ ]:
import numpy as np
import pandas as pd
import torch

from datasets import load_dataset, DatasetDict
from transformers import (
    AutoTokenizer,
    AutoModelForTokenClassification,
    DataCollatorForTokenClassification,
    TrainingArguments,
    Trainer,
    set_seed
)

from seqeval.metrics import (
    classification_report,
    f1_score,
    precision_score,
    recall_score
)

set_seed(42)

## 3. Dataset loading and exploration

We use the **CoNLL-2003** dataset from the `datasets` library.  
It contains three splits: `train`, `validation`, and `test`. Each example has:
- `tokens`: list of words in the sentence.
- `ner_tags`: list of integer labels (one per token).
The label IDs are later mapped to human-readable tags such as `B-PER`, `I-ORG`, `O`, etc.

The dataset follows the **BIO** labelling scheme:

- `B-XXX` marks the **Beginning** of an entity of type `XXX` (e.g. `B-PER` for the first token of a person name).
- `I-XXX` marks tokens **Inside** the same entity.
- `O` marks tokens that are **Outside** any named entity.


In [ ]:
from datasets import load_dataset
dataset = load_dataset("lhoestq/conll2003")
dataset


DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags'],
        num_rows: 3453
    })
})

In [ ]:
example = dataset["train"][0]
example

{'id': '0',
 'tokens': ['EU',
  'rejects',
  'German',
  'call',
  'to',
  'boycott',
  'British',
  'lamb',
  '.'],
 'pos_tags': [22, 42, 16, 21, 35, 37, 16, 21, 7],
 'chunk_tags': [11, 21, 11, 12, 21, 22, 11, 12, 0],
 'ner_tags': [3, 0, 7, 0, 0, 0, 7, 0, 0]}

In [ ]:
dataset["train"].features

{'id': Value('string'),
 'tokens': List(Value('string')),
 'pos_tags': List(Value('int64')),
 'chunk_tags': List(Value('int64')),
 'ner_tags': List(Value('int64'))}

In [ ]:
# CoNLL-2003 NER tag mapping (fixed)
label_list = [
    "O",        # 0
    "B-PER",    # 1
    "I-PER",    # 2
    "B-ORG",    # 3
    "I-ORG",    # 4
    "B-LOC",    # 5
    "I-LOC",    # 6
    "B-MISC",   # 7
    "I-MISC"    # 8
]

num_labels = len(label_list)

id2label = {i: label_list[i] for i in range(num_labels)}
label2id = {v: k for k, v in id2label.items()}

num_labels, id2label


(9,
 {0: 'O',
  1: 'B-PER',
  2: 'I-PER',
  3: 'B-ORG',
  4: 'I-ORG',
  5: 'B-LOC',
  6: 'I-LOC',
  7: 'B-MISC',
  8: 'I-MISC'})

In [ ]:
from collections import Counter

all_train_tags = [tag for sent in dataset["train"]["ner_tags"] for tag in sent]
Counter(all_train_tags)


Counter({3: 6321,
         0: 169578,
         7: 3438,
         1: 6600,
         2: 4528,
         5: 7140,
         4: 3704,
         8: 1155,
         6: 1157})

## 4. Preprocessing and tokenization

For token classification we must:
1. Tokenize each sentence while preserving the mapping from tokens (words) to subword tokens.
2. Align the original `ner_tags` with the new token indices.
3. Mark tokens that should be ignored in the loss (e.g. special tokens) with label `-100`.

I used the DistilBERT tokenizer with `is_split_into_words=True`, so that we can recover the word indices and align labels accordingly.

A subtle but important detail is how we handle **word-piece tokenization**.

Modern BERT-like models split rare or long words into sub-tokens (for example, `"Washington"` might become `"Wash"` and `"ington"`). However, labels in CoNLL-2003 are defined **per original word**, not per sub-token.

To reconcile this mismatch we:

- Propagate the original label to the **first sub-token** of each word.
- Mark all **subsequent sub-tokens** with label `-100` so that the loss function ignores them.
- Assign `-100` to special tokens such as `[CLS]` and `[SEP]` as well.

In [ ]:
from transformers import AutoTokenizer

model_checkpoint = "distilbert-base-uncased"
tokenizer = AutoTokenizer.from_pretrained(model_checkpoint, use_fast=True)


In [ ]:
def tokenize_and_align_labels(examples):
    tokenized = tokenizer(
        examples["tokens"],
        truncation=True,
        is_split_into_words=True
    )

    all_labels = examples["ner_tags"]
    new_labels = []

    for i in range(len(all_labels)):
        word_ids = tokenized.word_ids(batch_index=i)
        word_labels = all_labels[i]
        previous_word_idx = None
        label_ids = []

        for word_idx in word_ids:
            if word_idx is None:
                # special tokens, padding
                label_ids.append(-100)
            elif word_idx != previous_word_idx:
                # first sub-token of a word
                label_ids.append(word_labels[word_idx])
            else:
                # subsequent sub-tokens of the same word
                # we can repeat the label, or set -100 to ignore them
                label_ids.append(word_labels[word_idx])
            previous_word_idx = word_idx

        new_labels.append(label_ids)

    tokenized["labels"] = new_labels
    return tokenized


In [ ]:
tokenized_datasets = dataset.map(
    tokenize_and_align_labels,
    batched=True
)


Map:   0%|          | 0/3250 [00:00<?, ? examples/s]

In [ ]:
## smaller datasets to train in a faster way
small_train = tokenized_datasets["train"].shuffle(seed=42).select(range(5000))
small_valid = tokenized_datasets["validation"].shuffle(seed=42).select(range(1000))
small_test  = tokenized_datasets["test"].shuffle(seed=42).select(range(1000))

## If wants to use a smaller dataset just change to "small_train, small_valid and small_test" datasets
tokenized_datasets_dict = DatasetDict({
    "train": tokenized_datasets["train"],
    "validation": tokenized_datasets["validation"],
    "test": tokenized_datasets["test"]
})

tokenized_datasets_dict


DatasetDict({
    train: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags', 'input_ids', 'attention_mask', 'labels'],
        num_rows: 14041
    })
    validation: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags', 'input_ids', 'attention_mask', 'labels'],
        num_rows: 3250
    })
    test: Dataset({
        features: ['id', 'tokens', 'pos_tags', 'chunk_tags', 'ner_tags', 'input_ids', 'attention_mask', 'labels'],
        num_rows: 3453
    })
})

## 5. Model and training setup

We load a pre-trained **DistilBERT** model and adapt it for token classification by:
- Adding a classification head on top of the last hidden states.
- Setting the number of labels to match the NER tag set.

We also prepare a data collator that dynamically pads batches and keeps the `labels` aligned with the inputs.


In [ ]:
model = AutoModelForTokenClassification.from_pretrained(
    model_checkpoint,
    num_labels=num_labels
)

data_collator = DataCollatorForTokenClassification(tokenizer=tokenizer)


Some weights of DistilBertForTokenClassification were not initialized from the model checkpoint at distilbert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [ ]:
def align_predictions(predictions, label_ids):
    preds = np.argmax(predictions, axis=2)

    batch_size, seq_len = preds.shape
    out_preds, out_labels = [], []

    for i in range(batch_size):
        pred_i = []
        label_i = []
        for j in range(seq_len):
            if label_ids[i, j] != -100:
                pred_i.append(label_list[preds[i, j]])
                label_i.append(label_list[label_ids[i, j]])
        out_preds.append(pred_i)
        out_labels.append(label_i)
    return out_preds, out_labels


def compute_metrics(eval_pred):
    logits, labels = eval_pred
    preds, true_labels = align_predictions(logits, labels)

    return {
        "precision": precision_score(true_labels, preds),
        "recall": recall_score(true_labels, preds),
        "f1": f1_score(true_labels, preds)
    }


## 6. Fine-tuning DistilBERT for NER

We use the HuggingFace `Trainer` API to handle:
- The training loop and optimization.
- Periodic evaluation on the validation set.
- Saving the best model checkpoint according to the F1 score.

Hyperparameters are kept simple and standard for BERT-like models:
- Learning rate: 2e-5
- Batch size: 16
- Epochs: 3

These hyperparameters are intentionally conservative and commonly used for fine-tuning:

- A small learning rate (2e-5) prevents catastrophic forgetting of the pre-trained knowledge.
- A moderate batch size (16) balances gradient stability with memory constraints.
- Training for 3 epochs is often enough for CoNLL-2003; more epochs may slightly improve performance but also increase the risk of overfitting.

In [ ]:
batch_size = 16

training_args = TrainingArguments(
    output_dir="ner-distilbert-conll2003",
    learning_rate=2e-5,
    per_device_train_batch_size=batch_size,
    per_device_eval_batch_size=batch_size,
    report_to="none",
    num_train_epochs=3,
    weight_decay=0.01,
    logging_steps=100,
    metric_for_best_model="f1",
    fp16=torch.cuda.is_available()
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets_dict["train"],
    eval_dataset=tokenized_datasets_dict["validation"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)


/tmp/ipython-input-1922119026.py:16: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [ ]:
train_result = trainer.train()
train_result


TrainOutput(global_step=2634, training_loss=0.09038191173867738, metrics={'train_runtime': 340.4634, 'train_samples_per_second': 123.723, 'train_steps_per_second': 7.737, 'total_flos': 510122266253334.0, 'train_loss': 0.09038191173867738, 'epoch': 3.0})

## Interpretation of the Training

During fine-tuning, the model shows a stable and healthy learning curve. The training loss starts relatively high (around 0.69) and drops quickly within the first few hundred steps, indicating that the model adapts rapidly to the NER task once gradients begin flowing through the classification head.

Across the three epochs, the loss keeps decreasing smoothly, eventually reaching values below 0.04 in the later stages. This pattern suggests that the model is consistently improving without signs of divergence or oscillation. The small fluctuations near the end (e.g., between 0.03 and 0.06) are typical of fine-tuning with small batches and do not indicate instability.

The final average training loss (0.089) confirms that the model successfully learned the mapping between contextual token representations and the NER label space. Since NER is evaluated at the entity level rather than at the raw loss level, what matters is that the loss converged and remained low — which is exactly what we observe.

Overall, the training behaviour is characteristic of a well-tuned Transformer fine-tuning run: fast initial adaptation, steady convergence, and no overfitting symptoms at the loss level.

## 7. Evaluation

We evaluate the fine-tuned model on the **test split**, using NER-specific metrics:
- Precision
- Recall
- F1 score

The metrics are computed at the **entity level** (not per token) using `seqeval`.

In NER, token-level accuracy can be misleading: predicting all tokens as `O` can still give a high accuracy if entities are rare.

For that reason we focus on **entity-level** metrics, where a prediction is counted as correct only if:

- The entity type is correct (e.g. `PER` vs `LOC`), and
- The predicted span matches exactly the gold span (same start and end positions).

This makes the evaluation stricter but also more informative about the real usefulness of the model in downstream applications.


In [ ]:
test_metrics = trainer.evaluate(tokenized_datasets_dict["test"])
test_metrics


{'eval_loss': 0.12188401073217392,
 'eval_precision': 0.8903301886792453,
 'eval_recall': 0.8960360788037028,
 'eval_f1': 0.8931740210576127,
 'eval_runtime': 2.9487,
 'eval_samples_per_second': 1171.024,
 'eval_steps_per_second': 73.253,
 'epoch': 3.0}

## Interpretation of the Evaluation Results
The evaluation metrics show that the fine-tuned model generalizes well to the unseen test split. The F1-score of ~0.893 places the model in a strong performance range for DistilBERT on CoNLL-2003, especially considering that this is a lightweight architecture and was trained with a standard, non-tuned hyperparameter setup. The precision (0.890) and recall (0.896) are closely balanced, indicating that the model is not biased toward over-predicting or under-predicting entities.

The relatively low evaluation loss (0.1218) aligns with these strong metrics and confirms that the model is not overfitting: the validation behaviour reflects the same stability seen during training.

Overall, the model demonstrates solid entity-level performance, with high accuracy in both detecting and classifying named entities. This level of performance is sufficient for practical NER applications in English newswire, especially when speed and model size are relevant constraints.

In [ ]:
predictions, labels, _ = trainer.predict(tokenized_datasets_dict["test"])
preds, true_labels = align_predictions(predictions, labels)

print(classification_report(true_labels, preds))


              precision    recall  f1-score   support

         LOC       0.89      0.92      0.90      2124
        MISC       0.77      0.73      0.75       996
         ORG       0.86      0.88      0.87      2588
         PER       0.96      0.95      0.96      2718

   micro avg       0.89      0.90      0.89      8426
   macro avg       0.87      0.87      0.87      8426
weighted avg       0.89      0.90      0.89      8426



## Interpretation of the Prediction Report
The per-class performance reveals a model that is strong overall, but not uniformly strong across all entity types. The PER (person) class stands out with extremely high metrics (F1 ≈ 0.96), which is expected: person names tend to be frequent, well-structured, and relatively unambiguous in the CoNLL-2003 dataset. The model consistently identifies them with both high precision and high recall.

LOC (locations) and ORG (organizations) also show solid results, with F1-scores around 0.90 and 0.87 respectively. Errors here typically arise from cases where the distinction between a place and an institution is context-dependent—something even humans occasionally mislabel. The model’s recall for LOC (0.92) suggests it captures most location mentions, while maintaining good precision.

The MISC category is clearly the weakest, with an F1-score of ~0.75. The MISC class is heterogeneous, contains many rare entities, and often lacks strong contextual cues. Both precision (0.77) and recall (0.73) drop here, which pulls the macro average down.

Despite the class imbalance, the micro-averaged F1-score (~0.89) matches the earlier evaluation metrics, confirming consistent performance across evaluation methods. The weighted average mirrors the distribution of the dataset and reinforces the conclusion: the model performs well where the dataset is dense and clean, and degrades gracefully where entity definitions are noisy or underrepresented.

## 8. Qualitative analysis

To better understand the model behaviour, we inspect some predicted entities on individual sentences.
We compare:
- The original tokens.
- The predicted NER tags.

This helps to identify typical success and failure cases (e.g. rare entity names, ambiguous mentions).


In [ ]:
def predict_sentence(tokens):
    device = next(model.parameters()).device

    # 1) Encoding only to align words with sub-tokens (no tensors created)
    enc_align = tokenizer(
        tokens,
        is_split_into_words=True,
        truncation=True
    )
    word_ids = enc_align.word_ids()

    # 2) Encoding for the model forward pass (with PyTorch tensors)
    enc_model = tokenizer(
        tokens,
        is_split_into_words=True,
        return_tensors="pt",
        truncation=True
    )
    inputs = {k: v.to(device) for k, v in enc_model.items()}

    # 3) Forward
    with torch.no_grad():
        outputs = model(**inputs)

    logits = outputs.logits.detach().cpu()
    preds = logits.argmax(dim=2)[0].tolist()

    # 4) Reconstruct a single label per original word
    word_labels = []
    previous_word_idx = None

    for idx, word_idx in enumerate(word_ids):
        if word_idx is None or word_idx == previous_word_idx:
            continue
        label_id = preds[idx]
        word_labels.append((tokens[word_idx], label_list[label_id]))
        previous_word_idx = word_idx

    return word_labels


In [ ]:
example_tokens = dataset["test"][5]["tokens"]
gold_tags = [label_list[t] for t in dataset["test"][5]["ner_tags"]]

print("TOKENS:")
print(example_tokens)
print("\nGOLD:")
print(gold_tags)
print("\nPREDICTIONS:")
print(predict_sentence(example_tokens))


TOKENS:
['China', 'controlled', 'most', 'of', 'the', 'match', 'and', 'saw', 'several', 'chances', 'missed', 'until', 'the', '78th', 'minute', 'when', 'Uzbek', 'striker', 'Igor', 'Shkvyrin', 'took', 'advantage', 'of', 'a', 'misdirected', 'defensive', 'header', 'to', 'lob', 'the', 'ball', 'over', 'the', 'advancing', 'Chinese', 'keeper', 'and', 'into', 'an', 'empty', 'net', '.']

GOLD:
['B-LOC', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-MISC', 'O', 'B-PER', 'I-PER', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'O', 'B-MISC', 'O', 'O', 'O', 'O', 'O', 'O', 'O']

PREDICTIONS:
[('China', 'B-LOC'), ('controlled', 'O'), ('most', 'O'), ('of', 'O'), ('the', 'O'), ('match', 'O'), ('and', 'O'), ('saw', 'O'), ('several', 'O'), ('chances', 'O'), ('missed', 'O'), ('until', 'O'), ('the', 'O'), ('78th', 'O'), ('minute', 'O'), ('when', 'O'), ('Uzbek', 'B-MISC'), ('striker', 'O'), ('Igor', 'B-PER'), ('Shkvyrin', 'I-PER'), ('took', 'O'), ('advantage',

## Interpretation of the Qualitative Example

This example shows a clean, fully correct prediction from the model. All entity boundaries and all entity types match the gold annotations exactly:

China → B-LOC
Correctly identified as a location.

Uzbek → B-MISC
The model correctly assigns the miscellaneous category used in CoNLL-2003 for nationalities and demonyms.

Igor → B-PER and Shkvyrin → I-PER
The model reconstructs the multi-token person name perfectly, with the correct span boundary and BIO tags.

Chinese → B-MISC
Also correctly predicted according to the gold label, as another demonym-type token.

The remaining tokens are all correctly labeled as O, and no entity boundaries are missed or over-extended. This indicates that the model is not only learning the typical NER classes but is also internalizing subtler patterns such as demonyms and multi-token personal names. In this sentence, there are zero errors.

## 9. Conclusions and limitations

In this notebook we fine-tuned **DistilBERT** for **Named Entity Recognition** on the CoNLL-2003 dataset.

Main observations:
- The model reaches a reasonably high F1 score with a simple fine-tuning setup.
- DistilBERT, although smaller than BERT, is still strong enough for NER in English newswire.
- Most errors occur on rare entities, ambiguous mentions, or long multi-word names.

Limitations:
- We did not perform hyperparameter search or model comparison; the focus was on a clear, end-to-end fine-tuning pipeline.

Possible extensions:
- Tune hyperparameters.
- Try a larger model (e.g. `bert-base-cased`) or multilingual variants.
- Apply the same pipeline to a domain-specific NER dataset.
